### PR TITLE
Removes non combat click delay

### PR DESCRIPTION
**Pull request by Git-Nivrak**
# About the pull request
Basically #2130 but coded differently

# Explain why it's good for the game
#2130
# Testing Photographs and Procedure
<details>
<summary>Screenshots & Videos</summary>

Put screenshots and videos here with an empty line between the screenshots and the `<details>` tags.

</details>

# Changelog
:cl:
del: Removed non-combat click delay, interacting with items not on your person should be significantly smoother
balance: Clicking on the floor (open turfs) with an item will no longer cause a click delay
/:cl:

### CODE REVIEW

File: code/_onclick/click.dm [hunk 1]
```diff
@@ -144,12 +144,16 @@
 	SEND_SIGNAL(src, COMSIG_MOB_POST_CLICK, A, mods)
 	return
 
+#define ATTACKBY_HINT_NO_AFTERATTACK (1 << 0)
+#define ATTACKBY_HINT_UPDATE_NEXT_MOVE (1 << 1)
```
Comment (fira) [inline code/_onclick/click.dm:148]: Put that in define files not out there randomly, especially since they're used in multiple places

**Review by fira — DISMISSED** [review]

File: code/_onclick/item_attack.dm [hunk 2]
```diff
@@ -121,5 +122,5 @@
 		var/hit = H.attacked_by(src, user)
 		if (hit && hitsound)
 			playsound(loc, hitsound, 25, 1)
-		return hit
-	return TRUE
+		return (hit|ATTACKBY_HINT_UPDATE_NEXT_MOVE)
```
[the remark this answers is not in the document]
Comment (Git-Nivrak) [reply code/_onclick/item_attack.dm:125]: basically the first bit is whether to afterattack or not (which was the previous only purpose of the return value) so it allows using it this way

File: code/__DEFINES/_click.dm [hunk 3]
```diff
@@ -35,3 +35,7 @@
 
 //Pixel coordinates in screen_loc format ("[tile_x]:[pixel_x],[tile_y]:[pixel_y]")
 #define SCREEN_LOC "screen-loc"
+
+//Attackby hints
+#define ATTACKBY_HINT_NO_AFTERATTACK (1 << 0) // Makes it not call afterattack
+#define ATTACKBY_HINT_UPDATE_NEXT_MOVE (1 << 1) // Applies click delay
```
Comment (Drulikar) [inline code/__DEFINES/_click.dm:41]: You should instead autodoc the comments rather than just have them as comments

File: code/_onclick/click.dm [hunk 4]
```diff
@@ -146,16 +146,18 @@
...
 			// in case the attackby slept
 			if(!W)
 				UnarmedAttack(A, 1, mods)
 				return
 
-			W.afterattack(A, src, 1, mods)
+			afterattack_result = W.afterattack(A, src, 1, mods)
+
+		if(W.attack_speed && !src.contains(A) && (attackby_result & ATTACKBY_HINT_UPDATE_NEXT_MOVE) || (afterattack_result & ATTACKBY_HINT_UPDATE_NEXT_MOVE))
+			next_move += W.attack_speed
```
Comment (Drulikar) [inline code/_onclick/click.dm:160]: `next_move` used to be applied prior to a potential UnarmedAttack; now next_move is no longer modified if you return for an UnarmedAttack.

I have literally no idea what effects this has though; but this PR in general is a giant can of worms of possible niche interactions.

**Review by Drulikar — DISMISSED** [review]

File: code/_onclick/click.dm (hunk 4 above)
Comment (Git-Nivrak) [reply code/_onclick/click.dm:160]: Fixed (the UnarmedAttack next move at least, no idea about the rest of the bugs waiting to be found)

**Comment by Ansekishokuu** [comment]
issue with being able to destroy a pylon with no cooldown between attacks (bayonet slash in this instance, observed as a marine destroyed the pylon in a second)

and a very minor issue with health analyzers being able to scan very fast.

**Comment by Drulikar** [comment]
> issue with being able to destroy a pylon with no cooldown between attacks (bayonet slash in this instance, observed as a marine destroyed the pylon in a second)
> 
> and a very minor issue with health analyzers being able to scan very fast.

Pylon attacking should already be resolved. Health analyzers are not.

**Comment by Wintermote** [comment]
I'm not sure the full extent of this change, but the click delay on practically every single action that was introduced to prevent various edge case abuses but instead affected everything not JUST the edge cases, was one of the worst changes (or series of changes) ever made to this game.

You can't believe how relieved I am to finally see this being taken seriously. Getting rid of click delay makes a HUGE difference to how responsive the game feels to player actions like simply putting objects into a bag or webbing, throwing regular items, or putting stuff in disposals, and it's explicitly something *multiple* people I've tried to get into CM complained about.

To new players, it doesn't feel like it's some kind of balancing system. It feels like the game just drops half your inputs and is bad and laggy in general, and that frustration on top of trying to learn it just evaporated people's (already lukewarm) interest.

**Comment by Red-byte3D** [comment]
you can slash apcs very fast

**Comment by JackieEstegado** [comment]
https://github.com/user-attachments/assets/b09dc9b9-ddea-4f5f-bb5b-d39849222a82

Fireaxes

**Comment by irakligabunia** [comment]
^Above happens with breaching hammers too (B5 synth issue)

**Comment by JackieEstegado** [comment]
Gun Melee too.... shouldnt this be taken off of TM till the very bad things are found and fixed?

**Comment by harryob** [comment]
> Gun Melee too.... shouldnt this be taken off of TM till the very bad things are found and fixed?

you really should include more information than this on replication, because gun melee has a click delay as far as we can reproduce

and no, testmerges are for testing, as the name implies. it's kind of a ridiculous amount of work to test every items interaction with every map object

**Comment by JackieEstegado** [comment]
> > Gun Melee too.... shouldnt this be taken off of TM till the very bad things are found and fixed?
> 
> you really should include more information than this on replication, because gun melee has a click delay as far as we can reproduce
> 
> and no, testmerges are for testing, as the name implies. it's kind of a ridiculous amount of work to test every items interaction with every map object

After further investigation it appears that specifically clicking on burrowers there is no delay, any item ranged or melle

**Review by harryob — APPROVED** [review]

**Review by Drulikar — DISMISSED** [review]
See https://github.com/cmss13-devs/cmss13/issues/7883

**Comment by Blundir** [comment]
also you can spam hand/zip cuffs very fast now, spamming sound like crazy

**Review by Drulikar — CHANGES_REQUESTED** [review]
- [ ] "also you can spam hand/zip cuffs very fast now, spamming sound like crazy"
- [ ] https://github.com/cmss13-devs/cmss13/issues/7894
- [ ] https://github.com/cmss13-devs/cmss13/issues/7900

**Comment by Watermelon914** [comment]
> * [ ]  "also you can spam hand/zip cuffs very fast now, spamming sound like crazy"
> 
>     * [ ]  [hitting fire axe cabinet - no click delay #7894](https://github.com/cmss13-devs/cmss13/issues/7894)
> 
>     * [ ]  [no click delay - hitting secure glass door #7900](https://github.com/cmss13-devs/cmss13/issues/7900)

these can probably be fixed post merge, they don't seem super critical

File: code/_onclick/click.dm [hunk 5]
```diff
@@ -146,16 +146,20 @@
 
 /mob/proc/click_adjacent(atom/A, obj/item/W, mods)
 	if(W)
-		if(W.attack_speed && !src.contains(A)) //Not being worn or carried in the user's inventory somewhere, including internal storages.
-			next_move += W.attack_speed
-
-		if(!A.attackby(W, src, mods) && A && !QDELETED(A))
+		var/attackby_result = A.attackby(W, src, mods)
+		var/afterattack_result
+		if(!(attackby_result & ATTACKBY_HINT_NO_AFTERATTACK) && A && !QDELETED(A))
```
Comment (Drulikar) [inline code/_onclick/click.dm:151]: - If you're going to check if A is truthy and not deleted, you should do this first. If its not truthy it'd just be a runtime.
- Should be unnecessary to see if A is truthy and also test !QDELETED(A)

**Review by Drulikar — CHANGES_REQUESTED** [review]
Otherwise looks fine

**Review by Drulikar — APPROVED** [review]

**Pull request merged**